pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -172,60 +172,95 @@ func (m *ManagerImpl) GetNUMAHints(resource string, amount int) numamanager.Numa
-                    			if err != nil {
-                        			glog.Errorf("Err: Cannot convert deviceMask to int. %v")
-                        			break
-                    			}
-                    			if !duplicates[deviceMaskInt] {
-                        			duplicates[deviceMaskInt] = true
-                        			nm = append(nm, int64(deviceMaskInt))
-                    			}
-                	        }    
-                		//WIP for preferred     
-                		defaultMask[socket] = 1             
-                		break
+    	glog.Infof("[device-manager] Available devices for resource: %v", available)	
+	duplicate_frequency := make(map[int64]int64)
+	var socketValues []int64	
+ 	for availID := range available {
+		for _, device := range devices[resource] {
+			glog.Infof("[device-manager] AvailID: %v DeviceID: %v", availID, device)
+			if availID == device.ID {
+				socket := int64(device.Socket)
+				socketValues = append(socketValues, socket)	//slice of sockets with resource
+				if duplicate_frequency[socket] <= 1 {
+                    			duplicate_frequency[socket] += 1
+                		} else {
+                    			duplicate_frequency[socket] = 1
+                		}					
+			}						
+		}
+	}	
+	glog.Infof("[device-manager] Sockets with device: %v", socketValues)
+	
+	var count, devicesTotal int64 = 0, 0 	
+	var amount64 int64
+	amount64 = int64(amount)	
+	var availableSockets []int64
+	var notEmptySkts []int64
+	// Loop through device counts (duplicates of each socket) - return if resources not available on a socket
+	for socket, devicesPerSkt := range duplicate_frequency{
+ 		glog.Infof("[device-manager] Socket : %d , number of devices : %d", socket, devicesPerSkt)
+		devicesTotal = devicesTotal + devicesPerSkt
+		if devicesPerSkt !=0 {
+			notEmptySkts = append(notEmptySkts, socket)
+		}
+		if devicesPerSkt >= amount64 {
+			availableSockets = append(availableSockets, socket)
+			count++
+		}		
+ 	}
+	glog.Infof("[device-manager] Device Count Total: %v", devicesTotal)	
+
+	if count == 0 {
+		glog.Infof("[device-manager] No Socket has available resources - Preferred only")
+	}    
+	glog.Infof("[device-manager] Sockets with 1 or more available devices: %v", notEmptySkts)
+	glog.Infof("[device-manager] Sockets with enough available devices: %v", availableSockets)
+	
+	// Find the largest socket that is NOT empty
+        var largestSkt int64 = 0
+        for i := 0; i < len(notEmptySkts); i++ {
+                if largestSkt < notEmptySkts[i] {
+                        largestSkt = notEmptySkts[i]
+                }
+        }
+        glog.Infof("[device-manager] Largest socket that is NOT empty: %v", largestSkt)
+	
+	var socketMask []int64
+	fullMask:= make([][]int64,len(availableSockets))
+	var socket int64 = 0
+	var i int64
+	// Use largest socket to construct 2D slice (mask) for return to NUMA Manager
+	for j := 0; j < len(availableSockets); j++ {
```

Comment:
You could do something like 
for j, socket := range available sockets {
    //calculation on socket
}

Means you get the socket assigned without explicitly doing so in the loop